src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs

```diff
@@ -8,5 +13,68 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
         {
             return (BoundExpression)Visit(node);
         }
+
+        public override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.ConstantValueOpt, node.MethodOpt, node.ResultKind, left, right, type));
+        }
+
+        public override BoundNode VisitUserDefinedConditionalLogicalOperator(BoundUserDefinedConditionalLogicalOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.LogicalOperator, node.TrueOperator, node.FalseOperator, node.ResultKind, left, right, type));
+        }
+
+        private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator, Func<BoundBinaryOperatorBase, BoundExpression, BoundExpression, TypeSymbol, BoundBinaryOperatorBase> nodeUpdater)
+        {
+            // Use an explicit stack to avoid blowing the managed stack when visiting deeply-recursive
+            // binary nodes
+            var stack = ArrayBuilder<BoundBinaryOperatorBase>.GetInstance();
+            BoundBinaryOperatorBase currentBinary;
+            BoundExpression child = binaryOperator;
+
+            while (true)
+            {
+                if (child is BoundBinaryOperatorBase childBinary)
+                {
+                    currentBinary = childBinary;
+                    child = childBinary.Left;
+                    stack.Push(currentBinary);
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            Debug.Assert(stack.Count > 0);
+            currentBinary = null;
```

Comment:
>currentBinary = null; [](start = 12, length = 21)

This becomes redundant after the refactoring suggested above and can be removed.